src/sinks/aws_cloudwatch_logs/service.rs

```diff
@@ -0,0 +1,329 @@
+        let key = req.key;
+        let events = req
+            .events
+            .into_iter()
+            .map(|req| InputLogEvent {
+                message: req.message,
+                timestamp: req.timestamp,
+            })
+            .collect();
+        let svc = if let Some(svc) = &mut self.clients.get_mut(&key) {
+            svc.clone()
+        } else {
+            // Concurrency limit is 1 because we need token from previous request.
+            let svc = ServiceBuilder::new()
+                .buffer(1)
+                .concurrency_limit(1)
+                .rate_limit(
+                    self.request_settings.rate_limit_num,
+                    self.request_settings.rate_limit_duration,
+                )
+                .retry(
+                    self.request_settings
+                        .retry_policy(CloudwatchRetryLogic::new()),
+                )
+                .buffer(1)
+                .timeout(self.request_settings.timeout)
+                .service(CloudwatchLogsSvc::new(
+                    &self.config,
+                    &key,
+                    self.client.clone(),
+                ));
+
+            self.clients.insert(key, svc.clone());
+            svc
+        };
+
+        svc.oneshot(events)
+            .map_ok(move |_x| CloudwatchResponse {
+                events_count,
+                events_byte_size,
+            })
+            .map_err(Into::into)
+            .boxed()
+    }
+}
+
+impl CloudwatchLogsSvc {
+    pub fn new(
+        config: &CloudwatchLogsSinkConfig,
+        key: &CloudwatchKey,
+        client: CloudWatchLogsClient,
+    ) -> Self {
+        let group_name = key.group.clone();
+        let stream_name = key.stream.clone();
+
+        let create_missing_group = config.create_missing_group.unwrap_or(true);
+        let create_missing_stream = config.create_missing_stream.unwrap_or(true);
+
+        CloudwatchLogsSvc {
+            client,
+            stream_name,
+            group_name,
+            create_missing_group,
+            create_missing_stream,
+            token: None,
+            token_rx: None,
+        }
+    }
+
+    pub fn process_events(&self, events: Vec<InputLogEvent>) -> Vec<Vec<InputLogEvent>> {
+        let now = Utc::now();
+        // Acceptable range of Event timestamps.
+        let age_range = (now - Duration::days(14)).timestamp_millis()
+            ..(now + Duration::hours(2)).timestamp_millis();
+
+        let mut events = events
+            .into_iter()
+            .filter(|e| age_range.contains(&e.timestamp))
+            .collect::<Vec<_>>();
+
+        // Sort by timestamp
+        events.sort_by_key(|e| e.timestamp);
+
+        info!(message = "Sending events.", events = %events.len());
+
+        let mut event_batches = Vec::new();
+        if events.is_empty() {
+            // This should happen rarely.
+            event_batches.push(Vec::new());
```

Comment:
Do we _have _ to send an empty batch? 🤔

Reply:
I don't understand why we originally had this, but looking through the code I don't see a reason to keep it. I'm moving the date filtering logic to a higher level, which will make this branch impossible to reach anyway, so I'll just delete it.

Reply:
fixed in ad628ee0140d02cb601f09e1145031ea7bd15895